app/Services/Validation/ApplicationTimelineValidator.php

```diff
@@ -0,0 +1,420 @@
+        ];
+    }
+
+    /**
+     * Validator instance for the Basic Info step.
+     *
+     * @param JobApplication $application Job Application object.
+     *
+     * @return \Illuminate\Contracts\Validation\Validator
+     */
+    public function basicsValidator(JobApplication $application)
+    {
+        $application->loadMissing('job_poster');
+
+        // Check to see if this application is related to a Job Poster with
+        // a bilingual requirement.
+        $langRequirement = $application->job_poster->language_requirement_id;
+        $langTestRequirement = LanguageRequirement::where('name', 'bilingual_intermediate')
+            ->orWhere('name', 'bilingual_advanced')
+            ->pluck('id');
+
+        $validator = Validator::make($application->toArray(), $this->basicInfoSimpleRules());
+
+        // Conditionally check for 'language_test_confirmed' if the
+        // closure returns true.
+        $validator->sometimes(
+            'language_test_confirmed',
+            'required|boolean|accepted',
+            function ($input) use ($langRequirement, $langTestRequirement) {
+                return in_array($langRequirement, $langTestRequirement->toArray());
+            }
+        );
+
+        return $validator;
+    }
+
+    /**
+     * Helper function to return completeness for the Basic Info step.
+     *
+     * @param JobApplication $application Job Application object.
+     *
+     * @return boolean
+     */
+    public function basicsComplete(JobApplication $application)
+    {
+        $validator = $this->basicsValidator($application);
+        return $validator->passes();
+    }
+
+    /* End Step 1 --------------------------------------------------------------------------------------------------- */
+
+    /* Step 2 ------------------------------------------------------------------------------------------------------- */
+
+    /**
+     * Validator instance for the Experience step.
+     *
+     * @param JobApplication $application Job Application object.
+     *
+     * @return \Illuminate\Contracts\Validation\Validator
+     */
+    public function experienceValidator(JobApplication $application)
+    {
+        $application->loadMissing('job_poster', 'job_poster.criteria');
+
+        $essentialCriteriaType = CriteriaType::where('name', 'essential')->first()->id;
+
+        $jobCriteria = $application->job_poster->criteria;
+        $requiredCriteria = $jobCriteria->filter(function ($criterion) use ($essentialCriteriaType) {
+            return $criterion->criteria_type_id === $essentialCriteriaType;
+        });
+
+        $experienceSkillRules = array();
+
+        foreach ($requiredCriteria as $criterion) {
+            // Validate that every essential skill has a corresponding experience.
+            $experienceSkillRules[] = [
+                'skill_id' => new ContainsArrayWithAttributeRule('skill_id', $criterion->skill_id)
+            ];
```

Comment:
So experienceSkillRules will end up looking like:
```
[
  ['skill_id' => ContainsArrayWithAttributeRule('skill_id', 1)],
  ['skill_id' => ContainsArrayWithAttributeRule('skill_id', 2)],
]
```
Is this right? Do we need the nested arrays? 

Also, should each rule have the 'skill_id' key? If I understand the rule, this will check that the 'skill_id' field of the validated object is an array that _also_ has a 'skill_id' field.